### ISSUE TITLE
Improve two-way integration of Java and Ruby exceptions

### ISSUE THREAD

**Issue by headius** [jruby/jruby#4781]
For #4699, we removed the automatic initCause call for Java exceptions raised from Ruby. There were a few issues with that logic:

* Java does not automatically set cause for you, so there seemed to be a bit of a mismatch.
* The cause can only be set once, so we must check if it has been set already. Attempting to set it again raises an illegal argument exception from Java.
* The cause field is not volatile so you cannot check and set it safely across threads. It must be set under lock, and even then it is not guaranteed to be visible.
* Cause was getting set on unrelated exceptions just because they got re-raised with $! in flight.

In light of these problems, I think we should not reinstate automatic cause-setting for Java exceptions, but we do need a way to provide a cause. There's a few ways this might be done...I try to describe some below.

RubyException could hold a reference to its "container" RaiseException, so it can be re-raised without making a new one.

When providing the Java cause, this exception could be used, either through a new JRuby-specific method:
```ruby
ex.throwable
# or
ex.to_java(Throwable)
```
...or via some magic trickery during value conversion in JI, when we see a RubyException is being passed for a Throwable.

We have also discussed in the past building a parallel exception hierarchy for RaiseException. In this structure, all exception types that currently exist would still work, but for each Ruby exception there would be an equivalent Java exception of the same name. The two would be inseparable, like the proxy objects we use for JI. This would allow the Ruby exception to behave more naturally as a Java exception, and allow Java code to rescue Ruby exceptions by name (rather than the RaiseException dance we have today).

All this combined would improve our integration between Ruby and Java exceptions.

**Comment by headius** [comment]
I've started work on the parallel exception hierarchy on the parallel_exceptions branch.

**Comment by headius** [comment]
I have abandoned the old branch and have a new (working) branch in exceptions2.

The logic here works like this:

* Each RubyException knows how to construct its own RaiseException type, and holds a reference to it once created lazily.
* Throw a RubyException using the form `throw ex.getRaiseException()`.

At the moment this does not introduce any new Throwable types to the system. I need to work on cleaning up initialization, especially wrt backtraces.

**Comment by headius** [comment]
I have landed the parallel exception hierarchy. Will look into making it solve this issue.

**Comment by headius** [comment]
Ok, without any additional modification, we can now do the following to attach a Ruby exception as the cause for a Java exception:

```
$ jruby -e 'begin; raise "hello"; rescue; raise java.lang.RuntimeException.new($!.to_java); end'
Unhandled Java exception: java.lang.RuntimeException: org.jruby.exceptions.RuntimeError: (RuntimeError) hello
java.lang.RuntimeException: org.jruby.exceptions.RuntimeError: (RuntimeError) hello
         newInstance0 at sun/reflect/NativeConstructorAccessorImpl.java:-2
          newInstance at sun/reflect/NativeConstructorAccessorImpl.java:62
          newInstance at sun/reflect/DelegatingConstructorAccessorImpl.java:45
          newInstance at java/lang/reflect/Constructor.java:423
    newInstanceDirect at org/jruby/javasupport/JavaConstructor.java:278
                 call at org/jruby/java/invokers/ConstructorInvoker.java:87
                 call at org/jruby/java/invokers/ConstructorInvoker.java:176
         cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:355
                 call at org/jruby/runtime/callsite/CachingCallSite.java:180
                 call at org/jruby/java/proxies/ConcreteJavaProxy.java:56
         cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:355
                 call at org/jruby/runtime/callsite/CachingCallSite.java:180
          newInstance at org/jruby/RubyClass.java:1002
                 call at org/jruby/RubyClass$INVOKER$i$newInstance.gen:-1
                 call at org/jruby/internal/runtime/methods/DynamicMethod.java:202
                 call at org/jruby/java/proxies/ConcreteJavaProxy.java:158
         cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:344
                 call at org/jruby/runtime/callsite/CachingCallSite.java:170
     invokeOther6:new at -e:1
               <main> at -e:1
  invokeWithArguments at java/lang/invoke/MethodHandle.java:627
                 load at org/jruby/ir/Compiler.java:94
            runScript at org/jruby/Ruby.java:846
          runNormally at org/jruby/Ruby.java:765
          runNormally at org/jruby/Ruby.java:783
          runFromMain at org/jruby/Ruby.java:596
        doRunFromMain at org/jruby/Main.java:417
          internalRun at org/jruby/Main.java:305
                  run at org/jruby/Main.java:232
                 main at org/jruby/Main.java:204

Caused by:
org.jruby.exceptions.RuntimeError: (RuntimeError) hello
  <main> at -e:1
```

Remaining question for @kares @enebo: should it work automatically without the `to_java` call?

We'd need to set up all Exception classes to report Throwable as their "native" Java class, but I think that's all. They'd be eligible for passing to methods that take Throwable, and the toJava logic I added will retrieve the same throwable every time.

**Comment by kares** [comment]
> Remaining question for @kares @enebo: should it work automatically without the to_java call?

tempting but I am not sure if we're able to think through all the consequences in user-land.
... maybe a compromise would be to have a flag (on by default) to "degrade" back? (not that I want it)

**Comment by headius** [comment]
I believe all uncontended changes for this landed in #5065, which now provides real Java exceptions and two-way integration for most Ruby exceptions.
